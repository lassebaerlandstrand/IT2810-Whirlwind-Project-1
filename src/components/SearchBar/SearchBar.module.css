.searchBarContainer {
  display: flex;
  align-items: center;
  max-width: 12.5rem;
  border: 1px solid var(--inactive-border);
  border-radius: 0.4rem;
  background-color: #ffffff;
  padding: 0.5rem 0.75rem;
  box-sizing: border-box;

  &:hover {
    cursor: text;
  }
}

.searchInput {
  flex: 1;
  width: 100%;
  padding: 0 0.5rem;
  border: none;
  outline: none;
  font-size: 1rem;
  color: #333333;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
}

.searchInput::placeholder {
  color: #999999;
}

.searchIcon {
  display: flex;
  justify-content: center;
  align-items: center;
  color: #999999;
}

.label {
  font-size: 0.85rem;
}
