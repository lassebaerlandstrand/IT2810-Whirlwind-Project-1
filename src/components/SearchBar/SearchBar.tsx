import { IconSearch } from '@tabler/icons-react';
import { useEffect, useState } from 'react';
import styles from './SearchBar.module.css';

type SearchBarProps = {
  onSearch: (searchQuery: string) => void;
};

const SearchBar = ({ onSearch }: SearchBarProps) => {
  const [searchQuery, setSearchQuery] = useState<string>(() => sessionStorage.getItem('SearchQuery') || '');

  useEffect(() => {
    sessionStorage.setItem('SearchQuery', searchQuery);
    onSearch(searchQuery);
  }, [searchQuery, onSearch]);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearchQuery(event.target.value);
  };

  return (
    <label className={styles.label}>
      Search cities:
      <div className={styles.searchBarContainer} role="search">
        <span className={styles.searchIcon} data-testid="search-icon">
          {<IconSearch />}
        </span>
        <input
          type="text"
          placeholder="City Name..."
          value={searchQuery}
          onChange={handleChange}
          className={styles.searchInput}
        />
      </div>
    </label>
  );
};

export default SearchBar;
