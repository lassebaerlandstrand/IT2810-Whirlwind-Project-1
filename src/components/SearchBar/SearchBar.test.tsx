import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import { useState } from 'react';
import { describe, expect, it, vi } from 'vitest';
import type { Location } from '../../types/api-types';
import SearchBar from './SearchBar';

const dummyData: Location[] = [
  { city_name: 'New York', country_name: 'USA', latitude: '40.7128', longitude: '-74.006' },
  { city_name: 'London', country_name: 'UK', latitude: '51.5074', longitude: '-0.1278' },
  { city_name: 'Tokyo', country_name: 'Japan', latitude: '35.6895', longitude: '139.6917' },
];

describe('SearchBar Component', () => {
  beforeEach(() => {
    sessionStorage.clear();
  });

  it('should render the search input and icon', () => {
    const onSearchMock = vi.fn();
    render(<SearchBar onSearch={onSearchMock} />);

    const searchInput = screen.getByPlaceholderText('Search here...');
    const searchIcon = screen.getByTestId('search-icon');

    expect(searchInput).toBeInTheDocument();
    expect(searchIcon).toBeInTheDocument();
  });

  it('should call onSearch with the correct value when input changes', () => {
    const onSearchMock = vi.fn();
    render(<SearchBar onSearch={onSearchMock} />);

    const searchInput = screen.getByPlaceholderText('Search here...');
    fireEvent.change(searchInput, { target: { value: 'test' } });

    expect(onSearchMock).toHaveBeenCalledWith('test');
  });

  it('should call onSearch with an empty string when input is cleared', () => {
    const onSearchMock = vi.fn();
    render(<SearchBar onSearch={onSearchMock} />);

    const searchInput = screen.getByPlaceholderText('Search here...');
    fireEvent.change(searchInput, { target: { value: 'test' } });
    fireEvent.change(searchInput, { target: { value: '' } });

    expect(onSearchMock).toHaveBeenCalledWith('');
  });

  it('should save the selected search query in session storage', () => {
    expect(sessionStorage.getItem('SearchQuery')).toBe(null);

    const onSearchMock = vi.fn();
    render(<SearchBar onSearch={onSearchMock} />);

    const searchInput = screen.getByRole('textbox');
    fireEvent.change(searchInput, { target: { value: 'test' } });
    expect(sessionStorage.getItem('SearchQuery')).toBe('test');

    cleanup();
    render(<SearchBar onSearch={onSearchMock} />);

    expect(sessionStorage.getItem('SearchQuery')).toBe('test');
    const newSearchInput = screen.getByRole('textbox');
    expect(newSearchInput).toHaveValue('test');
  });
});

const SearchBarWithState = ({ onSearch }: { onSearch: (searchQuery: string) => void }) => {
  const [cities] = useState<Location[]>(dummyData);
  const [filteredCities, setFilteredCities] = useState(cities);

  const handleSearch = (searchQuery: string) => {
    if (searchQuery === '') {
      setFilteredCities(cities);
    } else {
      const filteredData = cities.filter((city) => city.city_name.toLowerCase().includes(searchQuery.toLowerCase()));
      setFilteredCities(filteredData);
    }
    onSearch(searchQuery);
  };

  return (
    <>
      <SearchBar onSearch={handleSearch} />
      <ul data-testid="filtered-cities">
        {filteredCities.map((city, index) => (
          <li key={index}>{city.city_name}</li>
        ))}
      </ul>
    </>
  );
};

describe('SearchBar Component with filtered cities', () => {
  it('should filter the cities based on the search query', () => {
    render(<SearchBarWithState onSearch={() => {}} />);

    const searchInput = screen.getByPlaceholderText('Search here...');
    fireEvent.change(searchInput, { target: { value: 'New York' } });

    const filteredCities = screen.getByTestId('filtered-cities'); // Assuming you have a test ID for the filtered cities list
    expect(filteredCities).toHaveTextContent('New York');
    expect(filteredCities).not.toHaveTextContent('London');
    expect(filteredCities).not.toHaveTextContent('Tokyo');
  });
});
