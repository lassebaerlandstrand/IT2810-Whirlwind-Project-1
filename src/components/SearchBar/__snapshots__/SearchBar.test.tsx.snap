// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBar Component > should match snapshot 1`] = `
<DocumentFragment>
  <label
    class="_label_ea9aeb"
  >
    Search cities:
    <div
      class="_searchBarContainer_ea9aeb"
      role="search"
    >
      <span
        class="_searchIcon_ea9aeb"
        data-testid="search-icon"
      >
        <svg
          class="tabler-icon tabler-icon-search "
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 10m-7 0a7 7 0 1 0 14 0a7 7 0 1 0 -14 0"
          />
          <path
            d="M21 21l-6 -6"
          />
        </svg>
      </span>
      <input
        class="_searchInput_ea9aeb"
        placeholder="City Name..."
        type="text"
        value=""
      />
    </div>
  </label>
</DocumentFragment>
`;
